src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.RenameDocumentUndoUnit.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.VisualStudio.OLE.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal partial class VisualStudioWorkspaceImpl
+    {
+        private class RenameDocumentUndoUnit : IOleUndoUnit
+        {
+            private readonly VisualStudioWorkspaceImpl _workspace;
+            private readonly string _fromName;
+            private readonly string _toName;
+            private readonly string _filePath;
+
+            public RenameDocumentUndoUnit(VisualStudioWorkspaceImpl workspace, string fromName, string toName, string filePath)
+            {
+                _workspace = workspace;
+                _fromName = fromName;
+                _toName = toName;
+                _filePath = filePath;
+            }
+
+            public void Do(IOleUndoManager pUndoManager)
+            {
+                // FirstOrDefault is okay because that's how the rename would have been
```

Comment:
Might be better to explain in isolation: "we only have to rename one document, since we will apply that to linked files too"?